pageserver/src/tenant/mgr.rs

```diff
@@ -1517,21 +1521,148 @@ impl TenantManager {
+                    .collect::<Vec<_>>()
+                    .await;
+                for layer in timeline_layers {
+                    let relative_path = layer
+                        .local_path()
+                        .strip_prefix(&parent_path)
+                        .context("Removing prefix from parent layer path")?;
+                    parent_layers.push(relative_path.to_owned());
+                }
+            }
+            (parent_timelines, parent_layers)
+        };
+        for child in child_shards {
+            let child_prefix = self.conf.tenant_path(&child);
+            let mut create_dirs = vec![child_prefix.clone()];
+            create_dirs.extend(
+                parent_timelines
+                    .iter()
+                    .map(|t| self.conf.timeline_path(&child, t)),
+            );
+            let layers_clone = parent_layers.clone();
+            let parent_path = parent_path.clone();
+
+            // Since we will do a large number of small filesystem metadata operations, batch them into
+            // spawn_blocking calls rather than doing each one as a tokio::fs round-trip.
+            let span = tracing::info_span!("link_layers", tenant_id=%child.tenant_id, shard_id=%child.shard_slug());
+            let jh = tokio::task::spawn_blocking(move || -> anyhow::Result<()> {
+                let _e = span.entered();
+                for dir in create_dirs {
+                    if let Err(e) = std::fs::create_dir_all(&dir) {
+                        // Ignore AlreadyExists errors, drop out on all other errors
+                        match e.kind() {
+                            std::io::ErrorKind::AlreadyExists => {}
+                            _ => {
+                                return Err(anyhow::anyhow!(e).context(format!("Creating {dir}")));
+                            }
+                        }
+                    }
+                }
+
+                for relative_layer in layers_clone {
+                    let parent_path = parent_path.join(&relative_layer);
+                    let child_path = child_prefix.join(&relative_layer);
+                    if let Err(e) = std::fs::hard_link(&parent_path, &child_path) {
+                        match e.kind() {
+                            std::io::ErrorKind::AlreadyExists => {}
+                            std::io::ErrorKind::NotFound => {
+                                tracing::info!(
+                                    "Layer {} not found during hard-linking, evicted during split?",
+                                    relative_layer
+                                );
+                            }
+                            _ => {
+                                return Err(anyhow::anyhow!(e).context(format!(
+                                    "Hard linking {relative_layer} into {child}"
+                                )))
+                            }
+                        }
+                    }
+                }
+
```

Comment:
Should there be a directory sync_all after we've finished hardlinking?

Reply:
It's not require for correctness, but as an optimization it makes sense, to avoid the risk that we could crash and restart with empty cacche. 0415f2873